packages/did-provider-ion/src/functions.ts

```diff
@@ -0,0 +1,284 @@
+    })
+    .filter((key) => !commitment || commitment === 'genesis' || !key.meta?.ion.commitment || key.meta?.ion.commitment === commitment)
+    .filter((key) => key.meta?.ion.relation === relation)
+    .flatMap((keys) => keys)
+}
+
+/**
+ * Ion/Sidetree only supports kid with a maximum length of 50 characters. This method truncates longer keys when create ION requests
+ *
+ * @param kid The Veramo kid
+ * @return A truncated kid if the input kid contained more than 50 characters
+ */
+export const truncateKidIfNeeded = (kid: string): string => {
+  const id = kid.substring(0, 50) // ION restricts the id to 50 chars. Ideally we can also provide kids for key creation in Veramo
+  if (id.length != kid.length) {
+    debug(`Key kid ${kid} has been truncated to 50 chars to support ION!`)
+  }
+  return id
+}
+
+/**
+ * Creates an Ion Public Key (Verification Method), used in ION request from a Veramo Key
+ * @param key The Veramo Key
+ * @param createKeyPurposes The verification relationships (Sidetree calls them purposes) to explicitly set. Used during key creation
+ * @return An Ion Public Key which can be used in Ion request objects
+ */
+export const toIonPublicKey = (key: ManagedKeyInfo, createKeyPurposes?: IonPublicKeyPurpose[]): IonPublicKeyModel => {
+  const purposes: IonPublicKeyPurpose[] = createKeyPurposes ? createKeyPurposes : key.meta?.ion?.purposes ? key.meta.ion.purposes : []
+  const publicKeyJwk = toIonPublicKeyJwk(key.publicKeyHex)
+  const id = truncateKidIfNeeded(key.kid)
+
+  return {
+    id,
+    type: 'EcdsaSecp256k1VerificationKey2019',
+    publicKeyJwk,
+    purposes,
+  }
+}
+
+/**
+ * Generates a random Private Hex Key for the specified key type
+ * @param type The key type
+ * @return The private key in Hex form
+ */
+export const generatePrivateKeyHex = (type: KeyType): string => {
+  let privateKeyHex: string
+
+  switch (type) {
+    case KeyType.Ed25519: {
+      const keyPairEd25519 = generateSigningKeyPair()
+      privateKeyHex = u8a.toString(keyPairEd25519.secretKey, 'base16')
+      break
+    }
+    case KeyType.Secp256k1: {
+      const privateBytes = randomBytes(32)
+      privateKeyHex = u8a.toString(privateBytes, 'base16')
+      break
+    }
+    default:
+      throw Error('not_supported: Key type not supported: ' + type)
+  }
+  return privateKeyHex
+}
+
+/**
+ * Create a Veramo Key entirely in Memory. It is not stored
+ *
+ * Didn't want to recreate the logic to extract the pub key for the different key types
+ * So let's create a temp in-mem kms to do it for us
+ *
+ * @param type
+ * @param privateKeyHex
+ * @param kid
+ * @param kms
+ * @param ionMeta
+ */
+export const tempMemoryKey = async (
+  type: KeyType.Ed25519 | KeyType.Secp256k1 | KeyType,
+  privateKeyHex: string,
+  kid: string,
+  kms: string,
+  ionMeta: IonKeyMetadata
+): Promise<IKey> => {
+  const tmpKey = (await new KeyManagementSystem(new MemoryPrivateKeyStore()).importKey({
```

Comment:
This is a good candidate for a utility function. It would then allow us to decouple this did-provider from the kms-local implementation.

Reply:
Add a ticket for it https://github.com/uport-project/veramo/issues/1021